sql/core/src/main/scala/org/apache/spark/sql/SQLContext.scala

```diff
@@ -129,19 +191,33 @@ class SQLContext(@transient val sparkContext: SparkContext)
   @transient
   protected[sql] lazy val optimizer: Optimizer = DefaultOptimizer
 
-  @transient
-  protected[sql] val ddlParser = new DDLParser(sqlParser.apply(_))
-
-  @transient
-  protected[sql] val sqlParser = {
-    val fallback = new catalyst.SqlParser
-    new SparkSQLParser(fallback(_))
+  protected[sql] def dialectCache = tlSession.get().dialectCache
+  protected[sql] def setDialectCache(d: SQLDialect) = {
+    tlSession.get().dialectCache = d
   }
 
-  protected[sql] def parseSql(sql: String): LogicalPlan = {
-    ddlParser(sql, false).getOrElse(sqlParser(sql))
+  protected[sql] def getSQLDialect(): SQLDialect = {
+    try {
+      val clazz = Utils.classForName(dialectClassName)
+      if (dialectCache == null || dialectCache.getClass != clazz) {
+        setDialectCache(clazz.newInstance().asInstanceOf[SQLDialect].initialize(this))
+      }
+
+      dialectCache
```

Comment:
This caching logic is pretty complicated and hard to follow.  Its not even caching the reflective part which is likely the most expensive part.  I would remote it and just construct dialects on demand unless you have a benchmark to show that is expensive.